vnext/ReactUWP/Views/Image/ReactImage.cpp

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include "ReactImage.h"
+
+#include <winrt/Windows.Web.Http.h>
+
+#include "unicode.h"
+
+namespace winrt {
+  using namespace Windows::Foundation;
+  using namespace Windows::Storage::Streams;
+  using namespace Windows::UI;
+  using namespace Windows::UI::Xaml::Media;
+  using namespace Windows::Web::Http;
+}
+
+namespace facebook {
+  using namespace facebook::react::unicode;
+}
+
+#if _MSC_VER <= 1913
+// VC 19 (2015-2017.6) cannot optimize co_await/cppwinrt usage
+#pragma optimize( "", off )
+#endif
+
+namespace react {
+  namespace uwp {
+
+    ReactImage::ReactImage()
+    {
+      m_brush = ReactImageBrush::Create();
+      this->Background(m_brush.as<winrt::XamlCompositionBrushBase>());
+    }
+
+    /*static*/ winrt::com_ptr<ReactImage> ReactImage::Create()
+    {
+      return winrt::make_self<ReactImage>();
+    }
+
+    winrt::Size ReactImage::ArrangeOverride(winrt::Size finalSize)
+    {
+      auto brush{ Background().as<ReactImageBrush>() };
+      brush->AvailableSize(finalSize);
+
+      return finalSize;
+    }
+
+    winrt::event_token ReactImage::OnLoadEnd(winrt::EventHandler<bool> const& handler)
+    {
+      return m_onLoadEndEvent.add(handler);
+    }
+
+    void ReactImage::OnLoadEnd(winrt::event_token const& token) noexcept
+    {
+      m_onLoadEndEvent.remove(token);
+    }
+
+    winrt::fire_and_forget ReactImage::Source(ImageSource source)
+    {
+      std::string uriString{ source.uri };
+
+      if (source.packagerAsset && uriString.find("file://") == 0)
+      {
+        uriString.replace(0, 7, "ms-appx:///Bundle/");
+      }
+
+      winrt::Uri uri{ facebook::utf8ToUtf16(uriString) };
+      winrt::hstring scheme{ uri.SchemeName() };
+      bool needsDownload = (scheme == L"http") || (scheme == L"https");
+
+      try
+      {
+        m_imageSource = source;
+
+        winrt::InMemoryRandomAccessStream memoryStream;
+        if (needsDownload)
+        {
+          memoryStream = co_await GetImageStreamAsync(source);
+
+          if (!memoryStream)
+          {
+            m_onLoadEndEvent(*this, false);
+          }
+        }
+
+        if (!needsDownload || memoryStream)
+        {
+          auto surface{ needsDownload ?
+            winrt::LoadedImageSurface::StartLoadFromStream(memoryStream) :
+            winrt::LoadedImageSurface::StartLoadFromUri(uri) };
+
+          surface.LoadCompleted({ this, &ReactImage::LoadedImageSurfaceHandler });
+
+          m_brush->Source(surface);
+        }
+      }
+      catch (winrt::hresult_error const&)
+      {
```

Comment:
Should this fire the load end event?
(like above when the memory stream fails?)